cpp/ql/lib/experimental/semmle/code/cpp/rangeanalysis/extensions/ConstantLShiftExprRange.qll

```diff
@@ -0,0 +1,152 @@
+private import cpp
+private import experimental.semmle.code.cpp.models.interfaces.SimpleRangeAnalysisExpr
+private import semmle.code.cpp.rangeanalysis.RangeAnalysisUtils
+
+float evaluateConstantExpr2(Expr e) {
+  result = e.getValue().toFloat()
+  or
+  // This handles when a constant value is put into a variable
+  // and the variable is used later
+  exists(SsaDefinition defn, StackVariable sv |
+    defn.getAUse(sv) = e and
+    result = defn.getDefiningValue(sv).getValue().toFloat()
+  )
+}
+
+// If the constant right operand is negative or is greater than or equal to the number of
+// bits in the left operands type, then the result is undefined (except on the IA-32
+// architecture where the shift value is masked with 0b00011111, but we can't
+// assume the architecture).
+bindingset[val]
+pragma[inline]
```

Comment:
A `bindingset` annotation causes the predicate to be inlined, so I don't understand why the `inline` pragma is also used? Is it because you really meant to say that both `val` and `l` should be bound already?

Reply:
It's been a while since I wrote this code, so I don't exactly remember how it ended up this way, but my guess is I started with `pragma[inline]` and after some refactoring needed the `bindingset`, and I didn't know that `bindingset` results in inlining the predicate. I'll remove the `pragma[inline]` here.